// @vexip-ui/hooks/dist/index.mjs
import { watch as fe, unref as D, getCurrentScope as Ye, onScopeDispose as Ue, ref as P, onMounted as _e, nextTick as ze, onBeforeUnmount as ve, computed as _, customRef as Sr, reactive as kt, readonly as kr, isRef as Ze, watchEffect as Rr, renderSlot as Tr } from "vue";
const oe = typeof window < "u";
var Rt;
oe && ((Rt = window == null ? void 0 : window.navigator) != null && Rt.userAgent) && /iP(ad|hone|od)/.test(window.navigator.userAgent);
function nt(e) {
  return e != null;
}
function F() {
}
const zr = Object.freeze({
  aliceblue: "f0f8ff",
  antiquewhite: "faebd7",
  aqua: "0ff",
  aquamarine: "7fffd4",
  azure: "f0ffff",
  beige: "f5f5dc",
  bisque: "ffe4c4",
  black: "000",
  blanchedalmond: "ffebcd",
  blue: "00f",
  blueviolet: "8a2be2",
  brown: "a52a2a",
  burlywood: "deb887",
  burntsienna: "ea7e5d",
  cadetblue: "5f9ea0",
  chartreuse: "7fff00",
  chocolate: "d2691e",
  coral: "ff7f50",
  cornflowerblue: "6495ed",
  cornsilk: "fff8dc",
  crimson: "dc143c",
  cyan: "0ff",
  darkblue: "00008b",
  darkcyan: "008b8b",
  darkgoldenrod: "b8860b",
  darkgray: "a9a9a9",
  darkgreen: "006400",
  darkgrey: "a9a9a9",
  darkkhaki: "bdb76b",
  darkmagenta: "8b008b",
  darkolivegreen: "556b2f",
  darkorange: "ff8c00",
  darkorchid: "9932cc",
  darkred: "8b0000",
  darksalmon: "e9967a",
  darkseagreen: "8fbc8f",
  darkslateblue: "483d8b",
  darkslategray: "2f4f4f",
  darkslategrey: "2f4f4f",
  darkturquoise: "00ced1",
  darkviolet: "9400d3",
  deeppink: "ff1493",
  deepskyblue: "00bfff",
  dimgray: "696969",
  dimgrey: "696969",
  dodgerblue: "1e90ff",
  firebrick: "b22222",
  floralwhite: "fffaf0",
  forestgreen: "228b22",
  fuchsia: "f0f",
  gainsboro: "dcdcdc",
  ghostwhite: "f8f8ff",
  gold: "ffd700",
  goldenrod: "daa520",
  gray: "808080",
  green: "008000",
  greenyellow: "adff2f",
  grey: "808080",
  honeydew: "f0fff0",
  hotpink: "ff69b4",
  indianred: "cd5c5c",
  indigo: "4b0082",
  ivory: "fffff0",
  khaki: "f0e68c",
  lavender: "e6e6fa",
  lavenderblush: "fff0f5",
  lawngreen: "7cfc00",
  lemonchiffon: "fffacd",
  lightblue: "add8e6",
  lightcoral: "f08080",
  lightcyan: "e0ffff",
  lightgoldenrodyellow: "fafad2",
  lightgray: "d3d3d3",
  lightgreen: "90ee90",
  lightgrey: "d3d3d3",
  lightpink: "ffb6c1",
  lightsalmon: "ffa07a",
  lightseagreen: "20b2aa",
  lightskyblue: "87cefa",
  lightslategray: "789",
  lightslategrey: "789",
  lightsteelblue: "b0c4de",
  lightyellow: "ffffe0",
  lime: "0f0",
  limegreen: "32cd32",
  linen: "faf0e6",
  magenta: "f0f",
  maroon: "800000",
  mediumaquamarine: "66cdaa",
  mediumblue: "0000cd",
  mediumorchid: "ba55d3",
  mediumpurple: "9370db",
  mediumseagreen: "3cb371",
  mediumslateblue: "7b68ee",
  mediumspringgreen: "00fa9a",
  mediumturquoise: "48d1cc",
  mediumvioletred: "c71585",
  midnightblue: "191970",
  mintcream: "f5fffa",
  mistyrose: "ffe4e1",
  moccasin: "ffe4b5",
  navajowhite: "ffdead",
  navy: "000080",
  oldlace: "fdf5e6",
  olive: "808000",
  olivedrab: "6b8e23",
  orange: "ffa500",
  orangered: "ff4500",
  orchid: "da70d6",
  palegoldenrod: "eee8aa",
  palegreen: "98fb98",
  paleturquoise: "afeeee",
  palevioletred: "db7093",
  papayawhip: "ffefd5",
  peachpuff: "ffdab9",
  peru: "cd853f",
  pink: "ffc0cb",
  plum: "dda0dd",
  powderblue: "b0e0e6",
  purple: "800080",
  rebeccapurple: "663399",
  red: "f00",
  rosybrown: "bc8f8f",
  royalblue: "4169e1",
  saddlebrown: "8b4513",
  salmon: "fa8072",
  sandybrown: "f4a460",
  seagreen: "2e8b57",
  seashell: "fff5ee",
  sienna: "a0522d",
  silver: "c0c0c0",
  skyblue: "87ceeb",
  slateblue: "6a5acd",
  slategray: "708090",
  slategrey: "708090",
  snow: "fffafa",
  springgreen: "00ff7f",
  steelblue: "4682b4",
  tan: "d2b48c",
  teal: "008080",
  thistle: "d8bfd8",
  tomato: "ff6347",
  turquoise: "40e0d0",
  violet: "ee82ee",
  wheat: "f5deb3",
  white: "fff",
  whitesmoke: "f5f5f5",
  yellow: "ff0",
  yellowgreen: "9acd32"
});
Object.freeze(new Set(Object.keys(zr)));
function Tt(e) {
  return e & -e;
}
function Br(e, t = 0) {
  const r = new Array(e + 1).fill(0);
  function n(i, s) {
    if (!(!s || i >= e))
      for (i += 1; i <= e; )
        r[i] += s, i += Tt(i);
  }
  function o(i = e) {
    if (i <= 0)
      return 0;
    i > e && (i = e);
    let s = i * t;
    for (; i > 0; )
      s += r[i], i -= Tt(i);
    return s;
  }
  function a(i) {
    return o(i + 1) - o(i);
  }
  function c(i) {
    let s = 0, u = e;
    for (; u > s; ) {
      const f = Math.floor((s + u) / 2), h = o(f);
      if (h > i) {
        u = f;
        continue;
      } else if (h < i) {
        if (s === f)
          return o(s + 1) <= i ? s + 1 : s;
        s = f;
      } else
        return f;
    }
    return s;
  }
  return { add: n, sum: o, get: a, boundIndex: c };
}
function Mr(e) {
  const t = parseFloat(e);
  return Number.isNaN(t) ? 0 : t;
}
const Ar = oe && ("ontouchstart" in window || Dr() > 0), Pr = Ar ? "pointerdown" : "click";
function Dr() {
  return typeof navigator < "u" && (navigator.maxTouchPoints || navigator.msMaxTouchPoints) || 0;
}
const J = /* @__PURE__ */ new Map();
function Cr(e) {
  J.has(e) || J.set(e, /* @__PURE__ */ new Set());
}
function Lr(e) {
  return J.get(e) ?? J.set(e, /* @__PURE__ */ new Set()).get(e);
}
function Wr(e, t) {
  if (typeof t == "string" && (t = [t]), Array.isArray(t))
    for (let r = 0, n = t.length; r < n; ++r) {
      const o = t[r];
      J.has(o) || J.set(o, /* @__PURE__ */ new Set()), J.get(o).add(e);
    }
}
function Fr(e, t) {
  var r;
  if (typeof t == "string" && (t = [t]), Array.isArray(t))
    for (let n = 0, o = t.length; n < o; ++n) {
      const a = t[n];
      J.has(a) && ((r = J.get(a)) == null || r.delete(e));
    }
}
function jr(e, t, r = window.Event) {
  const { type: n, bubbles: o = !1, cancelable: a = !1, ...c } = t;
  if (!nt(n) || n === "")
    return !1;
  let i;
  return nt(r) ? i = new r(n, { bubbles: o, cancelable: a }) : (i = document.createEvent("HTMLEvents"), i.initEvent(n, o, a)), Object.assign(i, c), e.dispatchEvent(i);
}
function Kt(e, t = 16) {
  if (typeof e != "function")
    return e;
  let r = Date.now(), n;
  return function(...o) {
    const a = Date.now(), c = r + t - a;
    clearTimeout(n), c <= 0 ? (e(...o), r = Date.now()) : n = setTimeout(() => {
      e(...o);
    }, t);
  };
}
const Se = /* @__PURE__ */ new Set(), Gt = /* @__PURE__ */ new WeakMap();
function $r() {
  Se.forEach((e) => {
    e(...Gt.get(e));
  }), Se.clear();
}
function Hr(e, ...t) {
  if (typeof e != "function")
    return e;
  Gt.set(e, t), !Se.has(e) && (Se.add(e), Se.size === 1 && requestAnimationFrame($r));
}
function Ir(e) {
  return Array.isArray(e) ? e : [e];
}
function Q(e, t, r, n) {
  if (!e)
    return F;
  let o = F;
  const a = fe(
    () => D(e),
    (i) => {
      o(), i && (i.addEventListener(t, r, n), o = () => {
        i.removeEventListener(t, r, n), o = F;
      });
    },
    { immediate: !0, flush: "post" }
  ), c = () => {
    a(), o();
  };
  return Ye() && Ue(c), c;
}
const he = "clickoutside";
Cr(he);
oe && document.addEventListener(
  Pr,
  (e) => {
    const t = e.target, r = e.composedPath && e.composedPath();
    Lr(he).forEach((n) => {
      n !== t && (r ? !r.includes(n) : !n.contains(t)) && (!n.__transferElement || n.__transferElement !== t && !n.__transferElement.contains(t)) && jr(n, { type: he });
    });
  },
  !0
);
function Bo(e, t = P(null)) {
  let r = F;
  const n = fe(
    () => D(t),
    (a) => {
      r(), a && (Wr(a, he), r = () => {
        Fr(a, he), r = F;
      });
    },
    { immediate: !0, flush: "post" }
  ), o = () => {
    n(), r();
  };
  return Ye() && Ue(o), Q(t, he, e), t;
}
function Mo(e = F, t = P(null)) {
  let r;
  return _e(() => {
    ze(() => {
      const n = Jt(t.value);
      n ? (r = new MutationObserver(() => {
        n.style.display !== "none" && (typeof e == "function" && e(), r == null || r.disconnect(), r = null);
      }), r.observe(n, {
        attributes: !0,
        childList: !0,
        characterData: !0,
        attributeFilter: ["style"]
      })) : typeof e == "function" && e();
    });
  }), ve(() => {
    r == null || r.disconnect(), r = null;
  }), t;
}
function Jt(e) {
  if (e) {
    let t = e.parentElement;
    for (; t && t !== document.body; ) {
      if (t.style.display === "none")
        return t;
      t = t.parentElement;
    }
  }
  return null;
}
function zt(e) {
  return (e == null ? void 0 : e.style.display) !== "none" ? !!Jt(e) : !0;
}
const Nr = [
  [
    "requestFullscreen",
    "exitFullscreen",
    "fullscreenElement",
    "fullscreenEnabled",
    "fullscreenchange",
    "fullscreenerror"
  ],
  // New WebKit
  [
    "webkitRequestFullscreen",
    "webkitExitFullscreen",
    "webkitFullscreenElement",
    "webkitFullscreenEnabled",
    "webkitfullscreenchange",
    "webkitfullscreenerror"
  ],
  // Old WebKit
  [
    "webkitRequestFullScreen",
    "webkitCancelFullScreen",
    "webkitCurrentFullScreenElement",
    "webkitCancelFullScreen",
    "webkitfullscreenchange",
    "webkitfullscreenerror"
  ],
  [
    "mozRequestFullScreen",
    "mozCancelFullScreen",
    "mozFullScreenElement",
    "mozFullScreenEnabled",
    "mozfullscreenchange",
    "mozfullscreenerror"
  ],
  [
    "msRequestFullscreen",
    "msExitFullscreen",
    "msFullscreenElement",
    "msFullscreenEnabled",
    "MSFullscreenChange",
    "MSFullscreenError"
  ]
];
let ge;
if (oe) {
  for (const e of Nr)
    if (e[1] in document) {
      ge = e;
      break;
    }
}
const ot = !!ge, qr = {
  supported: ot,
  full: _(() => !1),
  enter: F,
  exit: F,
  toggle: F
}, Qt = /* @__PURE__ */ new Set();
if (oe && ge) {
  const e = ge[2], t = ge[4];
  document.addEventListener(
    t,
    () => {
      const r = !!document[e];
      Qt.forEach((n) => {
        n.value = r;
      });
    },
    !1
  );
}
function Ao(e = P(null)) {
  if (!oe || !ot)
    return { ...qr };
  const [t, r, n] = ge, o = P(!1);
  async function a(s = !1) {
    await c(), e.value && (s || !document[n]) && (await e.value[t](), o.value = !0);
  }
  async function c(s = !1) {
    (s || document[n] && document[n] === e.value) && (await document[r](), o.value = !1);
  }
  async function i(s = !1) {
    o.value ? await c(s) : await a(s);
  }
  return Qt.add(o), Ye() && Ue(c), {
    supported: ot,
    target: e,
    full: _(() => o.value),
    enter: a,
    exit: c,
    toggle: i
  };
}
function Po(e = P(null)) {
  const t = P(!1);
  Q(e, "mouseenter", r), Q(e, "mouseleave", n);
  function r() {
    t.value = !0;
  }
  function n() {
    t.value = !1;
  }
  return { wrapper: e, isHover: t };
}
const Fe = /* @__PURE__ */ new WeakMap(), Xr = /\s+/g, Vr = /(px|%)$/;
function Do(e) {
  const t = e.target || P(null);
  if (!oe)
    return { target: t, disconnect: F };
  const { handler: r } = e;
  let n = e.root ?? document;
  const o = Ir(e.threshold || 0).join() || "0", a = g(e.rootMargin);
  Fe.has(n) || Fe.set(n, /* @__PURE__ */ new Map());
  const c = Fe.get(n);
  c.has(o) || c.set(o, /* @__PURE__ */ new Map());
  const i = c.get(o);
  i.has(a) || i.set(a, {
    ob: new IntersectionObserver(w, { ...e, rootMargin: a }),
    count: 0,
    handlers: /* @__PURE__ */ new WeakMap()
  });
  let s = i.get(a), { ob: u, handlers: f } = s, h = F;
  const y = fe(
    () => D(t),
    (l) => {
      h(), !(!l || !u) && (f.set(l, r), u.observe(l), s && s.count++, h = () => {
        u.unobserve(l), f.delete(l), s && s.count--, h = F;
      });
    },
    { immediate: !0, flush: "post" }
  );
  Ye() && Ue(d);
  function g(l) {
    if (!l || !l.trim())
      return "_";
    const p = l.trim().split(Xr, 4);
    p.length = 4;
    for (let b = 0; b < 4; ++b) {
      const x = p[b];
      p[b] = Vr.test(x) ? x : `${Mr(x)}px`;
    }
    return p.join(" ");
  }
  function w(l) {
    for (let p = 0, b = l.length; p < b; ++p) {
      const x = l[p], m = f.get(x.target);
      typeof m == "function" && m(x);
    }
  }
  function d() {
    y(), h(), s && (s.count <= 0 && (i.delete(a), i.size || (c.delete(o), c.size || Fe.delete(n))), s = void 0, u = void 0, f = void 0, n = void 0);
  }
  return { target: t, disconnect: d };
}
function Co() {
  const e = /* @__PURE__ */ new Set();
  function t(n) {
    return Sr((o, a) => {
      let c = n;
      const i = () => {
        n !== c && (c = n, a());
      };
      return {
        // eslint-disable-next-line no-sequences
        get: () => (o(), n),
        set: (s) => {
          s !== n && (n = s, e.add(i));
        }
      };
    });
  }
  function r() {
    for (const n of e)
      n();
    e.clear();
  }
  return { updateSet: e, manualRef: t, triggerUpdate: r };
}
const Yr = {
  ctrl: "control",
  command: "meta",
  cmd: "meta",
  option: "alt",
  up: "arrowup",
  down: "arrowdown",
  left: "arrowleft",
  right: "arrowright"
}, Ur = /[+_-]/, _r = /[+_-]/g, Kr = ["activeKeys", "resetAll"];
function Lo(e = {}) {
  const {
    autoReset: t = !0,
    capture: r = !1,
    passive: n = !0,
    onKeyDown: o = F,
    onKeyUp: a = F
  } = e, c = e.target || P(null), i = { ...Yr, ...e.aliasMap || {} }, s = kt(/* @__PURE__ */ new Set()), u = /* @__PURE__ */ new Set(), f = kt({ activeKeys: s, resetAll: g });
  function h(d, l) {
    d in f && (f[d] = l);
  }
  function y(d, l) {
    var m, O;
    const p = (m = d.key) == null ? void 0 : m.toLocaleLowerCase(), x = [(O = d.code) == null ? void 0 : O.toLocaleLowerCase(), p].filter(Boolean);
    for (const E of x)
      s[l ? "add" : "delete"](E), h(E, l);
    if (!l && p === "meta") {
      for (const E of u)
        s.delete(E), h(E, !1);
      u.clear();
    } else if (l && typeof d.getModifierState == "function" && d.getModifierState("Meta"))
      for (const E of [...s, ...x])
        u.add(E);
  }
  function g() {
    Object.keys(f).forEach((d) => {
      f[d] = !1;
    }), f.activeKeys = s, f.resetAll = g;
  }
  const w = new Proxy(f, {
    get(d, l, p) {
      if (typeof l != "string" || Kr.includes(l))
        return Reflect.get(d, l, p);
      if (l = l.toLocaleLowerCase(), l in i && (l = i[l]), !(l in f))
        if (Ur.test(l)) {
          const b = l.split(_r).map((x) => x.trim());
          f[l] = _(() => b.every((x) => D(w[x])));
        } else
          f[l] = P(s.has(l));
      return D(Reflect.get(d, l, p));
    }
  });
  return Q(
    c,
    "keydown",
    (d) => {
      y(d, !0), o(d, w);
    },
    { capture: r, passive: n }
  ), Q(
    c,
    "keyup",
    (d) => {
      y(d, !1), a(d, w);
    },
    { capture: r, passive: n }
  ), t && Q(c, "blur", g, { capture: r, passive: n }), { target: c, modifier: w };
}
function Wo(e) {
  const t = P(!1), r = () => t.value = !0;
  return _e(() => {
    e === "tick" ? ze(r) : e === "frame" ? requestAnimationFrame(r) : r();
  }), ve(() => {
    t.value = !1;
  }), { isMounted: kr(t) };
}
function et(e) {
  e.cancelable && (e.stopPropagation(), e.preventDefault());
}
function Fo(e) {
  const t = e.target || P(null), r = Ze(e.x) ? e.x : P(0), n = Ze(e.y) ? e.y : P(0), o = Ze(e.lazy) ? e.lazy : P(!1), { capture: a = !0, stopMouse: c = !0, stopTouch: i = !0 } = e, s = P(!1), u = {
    xStart: 0,
    yStart: 0,
    xEnd: 0,
    yEnd: 0,
    clientX: 0,
    clientY: 0,
    deltaX: 0,
    deltaY: 0,
    lazy: !1
  };
  let f = 0, h = 0;
  const y = Kt((p) => {
    var b;
    h < f || (l(p), u.lazy || (r.value = u.xEnd, n.value = u.yEnd), (b = e.onMove) == null || b.call(e, u, p));
  });
  function g(p) {
    var b;
    Object.assign(u, {
      xStart: r.value,
      yStart: n.value,
      xEnd: r.value,
      yEnd: n.value,
      clientX: p.clientX,
      clientY: p.clientY,
      lazy: o.value
    }), ((b = e.onStart) == null ? void 0 : b.call(e, u, p)) !== !1 && (document.addEventListener("pointermove", w, { capture: a }), document.addEventListener("pointerup", d, { capture: a }), h = f, s.value = !0);
  }
  function w(p) {
    et(p), y(p);
  }
  function d(p) {
    var b;
    document.removeEventListener("pointermove", w, { capture: a }), document.removeEventListener("pointerup", d, { capture: a }), l(p), u.lazy && (r.value = u.xEnd, n.value = u.yEnd), s.value = !1, ++f, (b = e.onEnd) == null || b.call(e, u, p);
  }
  function l(p) {
    const { clientX: b, clientY: x } = p, { xStart: m, yStart: O, clientX: E, clientY: R } = u, B = b - E, T = x - R;
    u.deltaX = B, u.deltaY = T, u.xEnd = m + B, u.yEnd = O + T;
  }
  return Q(t, "pointerdown", g, { capture: a }), c && Q(t, "mousedown", et, { capture: a }), i && Q(t, "touchstart", et, { capture: a }), {
    target: t,
    moving: _(() => s.value),
    x: r,
    y: n,
    lazy: o
  };
}
const Be = /* @__PURE__ */ new Set(), ne = /* @__PURE__ */ new Map();
ne.set("x", 0);
ne.set("y", 0);
function Gr(e) {
  const { pageX: t, pageY: r } = e;
  ne.set("x", t), ne.set("y", r), Be.forEach((n) => {
    n.x.value = t, n.y.value = r;
  });
}
const Zt = Kt(Gr);
function Jr(e) {
  !Be.size && window && (ne.set("x", 0), ne.set("y", 0), window.addEventListener("pointermove", Zt, { passive: !0 })), Be.add(e);
}
function Bt(e) {
  Be.delete(e), !Be.size && window && window.removeEventListener("pointermove", Zt);
}
function jo(e = {}) {
  const t = P(e.x ?? ne.get("x")), r = P(e.y ?? ne.get("y")), n = { x: t, y: r };
  return Jr(n), e.manualStop || ve(() => {
    Bt(n);
  }), { ...n, unregister: () => Bt(n) };
}
var j = "top", N = "bottom", q = "right", $ = "left", ct = "auto", De = [j, N, q, $], be = "start", Me = "end", Qr = "clippingParents", er = "viewport", Oe = "popper", Zr = "reference", Mt = /* @__PURE__ */ De.reduce(function(e, t) {
  return e.concat([t + "-" + be, t + "-" + Me]);
}, []), tr = /* @__PURE__ */ [].concat(De, [ct]).reduce(function(e, t) {
  return e.concat([t, t + "-" + be, t + "-" + Me]);
}, []), en = "beforeRead", tn = "read", rn = "afterRead", nn = "beforeMain", on = "main", an = "afterMain", sn = "beforeWrite", cn = "write", fn = "afterWrite", un = [en, tn, rn, nn, on, an, sn, cn, fn];
function G(e) {
  return e ? (e.nodeName || "").toLowerCase() : null;
}
function H(e) {
  if (e == null)
    return window;
  if (e.toString() !== "[object Window]") {
    var t = e.ownerDocument;
    return t && t.defaultView || window;
  }
  return e;
}
function pe(e) {
  var t = H(e).Element;
  return e instanceof t || e instanceof Element;
}
function I(e) {
  var t = H(e).HTMLElement;
  return e instanceof t || e instanceof HTMLElement;
}
function ft(e) {
  if (typeof ShadowRoot > "u")
    return !1;
  var t = H(e).ShadowRoot;
  return e instanceof t || e instanceof ShadowRoot;
}
function ln(e) {
  var t = e.state;
  Object.keys(t.elements).forEach(function(r) {
    var n = t.styles[r] || {}, o = t.attributes[r] || {}, a = t.elements[r];
    !I(a) || !G(a) || (Object.assign(a.style, n), Object.keys(o).forEach(function(c) {
      var i = o[c];
      i === !1 ? a.removeAttribute(c) : a.setAttribute(c, i === !0 ? "" : i);
    }));
  });
}
function dn(e) {
  var t = e.state, r = {
    popper: {
      position: t.options.strategy,
      left: "0",
      top: "0",
      margin: "0"
    },
    arrow: {
      position: "absolute"
    },
    reference: {}
  };
  return Object.assign(t.elements.popper.style, r.popper), t.styles = r, t.elements.arrow && Object.assign(t.elements.arrow.style, r.arrow), function() {
    Object.keys(t.elements).forEach(function(n) {
      var o = t.elements[n], a = t.attributes[n] || {}, c = Object.keys(t.styles.hasOwnProperty(n) ? t.styles[n] : r[n]), i = c.reduce(function(s, u) {
        return s[u] = "", s;
      }, {});
      !I(o) || !G(o) || (Object.assign(o.style, i), Object.keys(a).forEach(function(s) {
        o.removeAttribute(s);
      }));
    });
  };
}
const pn = {
  name: "applyStyles",
  enabled: !0,
  phase: "write",
  fn: ln,
  effect: dn,
  requires: ["computeStyles"]
};
function K(e) {
  return e.split("-")[0];
}
var ue = Math.max, Xe = Math.min, ye = Math.round;
function it() {
  var e = navigator.userAgentData;
  return e != null && e.brands && Array.isArray(e.brands) ? e.brands.map(function(t) {
    return t.brand + "/" + t.version;
  }).join(" ") : navigator.userAgent;
}
function rr() {
  return !/^((?!chrome|android).)*safari/i.test(it());
}
function we(e, t, r) {
  t === void 0 && (t = !1), r === void 0 && (r = !1);
  var n = e.getBoundingClientRect(), o = 1, a = 1;
  t && I(e) && (o = e.offsetWidth > 0 && ye(n.width) / e.offsetWidth || 1, a = e.offsetHeight > 0 && ye(n.height) / e.offsetHeight || 1);
  var c = pe(e) ? H(e) : window, i = c.visualViewport, s = !rr() && r, u = (n.left + (s && i ? i.offsetLeft : 0)) / o, f = (n.top + (s && i ? i.offsetTop : 0)) / a, h = n.width / o, y = n.height / a;
  return {
    width: h,
    height: y,
    top: f,
    right: u + h,
    bottom: f + y,
    left: u,
    x: u,
    y: f
  };
}
function ut(e) {
  var t = we(e), r = e.offsetWidth, n = e.offsetHeight;
  return Math.abs(t.width - r) <= 1 && (r = t.width), Math.abs(t.height - n) <= 1 && (n = t.height), {
    x: e.offsetLeft,
    y: e.offsetTop,
    width: r,
    height: n
  };
}
function nr(e, t) {
  var r = t.getRootNode && t.getRootNode();
  if (e.contains(t))
    return !0;
  if (r && ft(r)) {
    var n = t;
    do {
      if (n && e.isSameNode(n))
        return !0;
      n = n.parentNode || n.host;
    } while (n);
  }
  return !1;
}
function Z(e) {
  return H(e).getComputedStyle(e);
}
function vn(e) {
  return ["table", "td", "th"].indexOf(G(e)) >= 0;
}
function ie(e) {
  return ((pe(e) ? e.ownerDocument : (
    // $FlowFixMe[prop-missing]
    e.document
  )) || window.document).documentElement;
}
function Ke(e) {
  return G(e) === "html" ? e : (
    // this is a quicker (but less type safe) way to save quite some bytes from the bundle
    // $FlowFixMe[incompatible-return]
    // $FlowFixMe[prop-missing]
    e.assignedSlot || // step into the shadow DOM of the parent of a slotted node
    e.parentNode || // DOM Element detected
    (ft(e) ? e.host : null) || // ShadowRoot detected
    // $FlowFixMe[incompatible-call]: HTMLElement is a Node
    ie(e)
  );
}
function At(e) {
  return !I(e) || // https://github.com/popperjs/popper-core/issues/837
  Z(e).position === "fixed" ? null : e.offsetParent;
}
function hn(e) {
  var t = /firefox/i.test(it()), r = /Trident/i.test(it());
  if (r && I(e)) {
    var n = Z(e);
    if (n.position === "fixed")
      return null;
  }
  var o = Ke(e);
  for (ft(o) && (o = o.host); I(o) && ["html", "body"].indexOf(G(o)) < 0; ) {
    var a = Z(o);
    if (a.transform !== "none" || a.perspective !== "none" || a.contain === "paint" || ["transform", "perspective"].indexOf(a.willChange) !== -1 || t && a.willChange === "filter" || t && a.filter && a.filter !== "none")
      return o;
    o = o.parentNode;
  }
  return null;
}
function Ce(e) {
  for (var t = H(e), r = At(e); r && vn(r) && Z(r).position === "static"; )
    r = At(r);
  return r && (G(r) === "html" || G(r) === "body" && Z(r).position === "static") ? t : r || hn(e) || t;
}
function lt(e) {
  return ["top", "bottom"].indexOf(e) >= 0 ? "x" : "y";
}
function ke(e, t, r) {
  return ue(e, Xe(t, r));
}
function gn(e, t, r) {
  var n = ke(e, t, r);
  return n > r ? r : n;
}
function or() {
  return {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
  };
}
function ir(e) {
  return Object.assign({}, or(), e);
}
function ar(e, t) {
  return t.reduce(function(r, n) {
    return r[n] = e, r;
  }, {});
}
var mn = function(t, r) {
  return t = typeof t == "function" ? t(Object.assign({}, r.rects, {
    placement: r.placement
  })) : t, ir(typeof t != "number" ? t : ar(t, De));
};
function bn(e) {
  var t, r = e.state, n = e.name, o = e.options, a = r.elements.arrow, c = r.modifiersData.popperOffsets, i = K(r.placement), s = lt(i), u = [$, q].indexOf(i) >= 0, f = u ? "height" : "width";
  if (!(!a || !c)) {
    var h = mn(o.padding, r), y = ut(a), g = s === "y" ? j : $, w = s === "y" ? N : q, d = r.rects.reference[f] + r.rects.reference[s] - c[s] - r.rects.popper[f], l = c[s] - r.rects.reference[s], p = Ce(a), b = p ? s === "y" ? p.clientHeight || 0 : p.clientWidth || 0 : 0, x = d / 2 - l / 2, m = h[g], O = b - y[f] - h[w], E = b / 2 - y[f] / 2 + x, R = ke(m, E, O), B = s;
    r.modifiersData[n] = (t = {}, t[B] = R, t.centerOffset = R - E, t);
  }
}
function yn(e) {
  var t = e.state, r = e.options, n = r.element, o = n === void 0 ? "[data-popper-arrow]" : n;
  o != null && (typeof o == "string" && (o = t.elements.popper.querySelector(o), !o) || nr(t.elements.popper, o) && (t.elements.arrow = o));
}
const wn = {
  name: "arrow",
  enabled: !0,
  phase: "main",
  fn: bn,
  effect: yn,
  requires: ["popperOffsets"],
  requiresIfExists: ["preventOverflow"]
};
function xe(e) {
  return e.split("-")[1];
}
var xn = {
  top: "auto",
  right: "auto",
  bottom: "auto",
  left: "auto"
};
function En(e, t) {
  var r = e.x, n = e.y, o = t.devicePixelRatio || 1;
  return {
    x: ye(r * o) / o || 0,
    y: ye(n * o) / o || 0
  };
}
function Pt(e) {
  var t, r = e.popper, n = e.popperRect, o = e.placement, a = e.variation, c = e.offsets, i = e.position, s = e.gpuAcceleration, u = e.adaptive, f = e.roundOffsets, h = e.isFixed, y = c.x, g = y === void 0 ? 0 : y, w = c.y, d = w === void 0 ? 0 : w, l = typeof f == "function" ? f({
    x: g,
    y: d
  }) : {
    x: g,
    y: d
  };
  g = l.x, d = l.y;
  var p = c.hasOwnProperty("x"), b = c.hasOwnProperty("y"), x = $, m = j, O = window;
  if (u) {
    var E = Ce(r), R = "clientHeight", B = "clientWidth";
    if (E === H(r) && (E = ie(r), Z(E).position !== "static" && i === "absolute" && (R = "scrollHeight", B = "scrollWidth")), E = E, o === j || (o === $ || o === q) && a === Me) {
      m = N;
      var T = h && E === O && O.visualViewport ? O.visualViewport.height : (
        // $FlowFixMe[prop-missing]
        E[R]
      );
      d -= T - n.height, d *= s ? 1 : -1;
    }
    if (o === $ || (o === j || o === N) && a === Me) {
      x = q;
      var A = h && E === O && O.visualViewport ? O.visualViewport.width : (
        // $FlowFixMe[prop-missing]
        E[B]
      );
      g -= A - n.width, g *= s ? 1 : -1;
    }
  }
  var C = Object.assign({
    position: i
  }, u && xn), L = f === !0 ? En({
    x: g,
    y: d
  }, H(r)) : {
    x: g,
    y: d
  };
  if (g = L.x, d = L.y, s) {
    var v;
    return Object.assign({}, C, (v = {}, v[m] = b ? "0" : "", v[x] = p ? "0" : "", v.transform = (O.devicePixelRatio || 1) <= 1 ? "translate(" + g + "px, " + d + "px)" : "translate3d(" + g + "px, " + d + "px, 0)", v));
  }
  return Object.assign({}, C, (t = {}, t[m] = b ? d + "px" : "", t[x] = p ? g + "px" : "", t.transform = "", t));
}
function On(e) {
  var t = e.state, r = e.options, n = r.gpuAcceleration, o = n === void 0 ? !0 : n, a = r.adaptive, c = a === void 0 ? !0 : a, i = r.roundOffsets, s = i === void 0 ? !0 : i, u = {
    placement: K(t.placement),
    variation: xe(t.placement),
    popper: t.elements.popper,
    popperRect: t.rects.popper,
    gpuAcceleration: o,
    isFixed: t.options.strategy === "fixed"
  };
  t.modifiersData.popperOffsets != null && (t.styles.popper = Object.assign({}, t.styles.popper, Pt(Object.assign({}, u, {
    offsets: t.modifiersData.popperOffsets,
    position: t.options.strategy,
    adaptive: c,
    roundOffsets: s
  })))), t.modifiersData.arrow != null && (t.styles.arrow = Object.assign({}, t.styles.arrow, Pt(Object.assign({}, u, {
    offsets: t.modifiersData.arrow,
    position: "absolute",
    adaptive: !1,
    roundOffsets: s
  })))), t.attributes.popper = Object.assign({}, t.attributes.popper, {
    "data-popper-placement": t.placement
  });
}
const Sn = {
  name: "computeStyles",
  enabled: !0,
  phase: "beforeWrite",
  fn: On,
  data: {}
};
var je = {
  passive: !0
};
function kn(e) {
  var t = e.state, r = e.instance, n = e.options, o = n.scroll, a = o === void 0 ? !0 : o, c = n.resize, i = c === void 0 ? !0 : c, s = H(t.elements.popper), u = [].concat(t.scrollParents.reference, t.scrollParents.popper);
  return a && u.forEach(function(f) {
    f.addEventListener("scroll", r.update, je);
  }), i && s.addEventListener("resize", r.update, je), function() {
    a && u.forEach(function(f) {
      f.removeEventListener("scroll", r.update, je);
    }), i && s.removeEventListener("resize", r.update, je);
  };
}
const Rn = {
  name: "eventListeners",
  enabled: !0,
  phase: "write",
  fn: function() {
  },
  effect: kn,
  data: {}
};
var Tn = {
  left: "right",
  right: "left",
  bottom: "top",
  top: "bottom"
};
function Ne(e) {
  return e.replace(/left|right|bottom|top/g, function(t) {
    return Tn[t];
  });
}
var zn = {
  start: "end",
  end: "start"
};
function Dt(e) {
  return e.replace(/start|end/g, function(t) {
    return zn[t];
  });
}
function dt(e) {
  var t = H(e), r = t.pageXOffset, n = t.pageYOffset;
  return {
    scrollLeft: r,
    scrollTop: n
  };
}
function pt(e) {
  return we(ie(e)).left + dt(e).scrollLeft;
}
function Bn(e, t) {
  var r = H(e), n = ie(e), o = r.visualViewport, a = n.clientWidth, c = n.clientHeight, i = 0, s = 0;
  if (o) {
    a = o.width, c = o.height;
    var u = rr();
    (u || !u && t === "fixed") && (i = o.offsetLeft, s = o.offsetTop);
  }
  return {
    width: a,
    height: c,
    x: i + pt(e),
    y: s
  };
}
function Mn(e) {
  var t, r = ie(e), n = dt(e), o = (t = e.ownerDocument) == null ? void 0 : t.body, a = ue(r.scrollWidth, r.clientWidth, o ? o.scrollWidth : 0, o ? o.clientWidth : 0), c = ue(r.scrollHeight, r.clientHeight, o ? o.scrollHeight : 0, o ? o.clientHeight : 0), i = -n.scrollLeft + pt(e), s = -n.scrollTop;
  return Z(o || r).direction === "rtl" && (i += ue(r.clientWidth, o ? o.clientWidth : 0) - a), {
    width: a,
    height: c,
    x: i,
    y: s
  };
}
function vt(e) {
  var t = Z(e), r = t.overflow, n = t.overflowX, o = t.overflowY;
  return /auto|scroll|overlay|hidden/.test(r + o + n);
}
function sr(e) {
  return ["html", "body", "#document"].indexOf(G(e)) >= 0 ? e.ownerDocument.body : I(e) && vt(e) ? e : sr(Ke(e));
}
function Re(e, t) {
  var r;
  t === void 0 && (t = []);
  var n = sr(e), o = n === ((r = e.ownerDocument) == null ? void 0 : r.body), a = H(n), c = o ? [a].concat(a.visualViewport || [], vt(n) ? n : []) : n, i = t.concat(c);
  return o ? i : (
    // $FlowFixMe[incompatible-call]: isBody tells us target will be an HTMLElement here
    i.concat(Re(Ke(c)))
  );
}
function at(e) {
  return Object.assign({}, e, {
    left: e.x,
    top: e.y,
    right: e.x + e.width,
    bottom: e.y + e.height
  });
}
function An(e, t) {
  var r = we(e, !1, t === "fixed");
  return r.top = r.top + e.clientTop, r.left = r.left + e.clientLeft, r.bottom = r.top + e.clientHeight, r.right = r.left + e.clientWidth, r.width = e.clientWidth, r.height = e.clientHeight, r.x = r.left, r.y = r.top, r;
}
function Ct(e, t, r) {
  return t === er ? at(Bn(e, r)) : pe(t) ? An(t, r) : at(Mn(ie(e)));
}
function Pn(e) {
  var t = Re(Ke(e)), r = ["absolute", "fixed"].indexOf(Z(e).position) >= 0, n = r && I(e) ? Ce(e) : e;
  return pe(n) ? t.filter(function(o) {
    return pe(o) && nr(o, n) && G(o) !== "body";
  }) : [];
}
function Dn(e, t, r, n) {
  var o = t === "clippingParents" ? Pn(e) : [].concat(t), a = [].concat(o, [r]), c = a[0], i = a.reduce(function(s, u) {
    var f = Ct(e, u, n);
    return s.top = ue(f.top, s.top), s.right = Xe(f.right, s.right), s.bottom = Xe(f.bottom, s.bottom), s.left = ue(f.left, s.left), s;
  }, Ct(e, c, n));
  return i.width = i.right - i.left, i.height = i.bottom - i.top, i.x = i.left, i.y = i.top, i;
}
function cr(e) {
  var t = e.reference, r = e.element, n = e.placement, o = n ? K(n) : null, a = n ? xe(n) : null, c = t.x + t.width / 2 - r.width / 2, i = t.y + t.height / 2 - r.height / 2, s;
  switch (o) {
    case j:
      s = {
        x: c,
        y: t.y - r.height
      };
      break;
    case N:
      s = {
        x: c,
        y: t.y + t.height
      };
      break;
    case q:
      s = {
        x: t.x + t.width,
        y: i
      };
      break;
    case $:
      s = {
        x: t.x - r.width,
        y: i
      };
      break;
    default:
      s = {
        x: t.x,
        y: t.y
      };
  }
  var u = o ? lt(o) : null;
  if (u != null) {
    var f = u === "y" ? "height" : "width";
    switch (a) {
      case be:
        s[u] = s[u] - (t[f] / 2 - r[f] / 2);
        break;
      case Me:
        s[u] = s[u] + (t[f] / 2 - r[f] / 2);
        break;
    }
  }
  return s;
}
function Ae(e, t) {
  t === void 0 && (t = {});
  var r = t, n = r.placement, o = n === void 0 ? e.placement : n, a = r.strategy, c = a === void 0 ? e.strategy : a, i = r.boundary, s = i === void 0 ? Qr : i, u = r.rootBoundary, f = u === void 0 ? er : u, h = r.elementContext, y = h === void 0 ? Oe : h, g = r.altBoundary, w = g === void 0 ? !1 : g, d = r.padding, l = d === void 0 ? 0 : d, p = ir(typeof l != "number" ? l : ar(l, De)), b = y === Oe ? Zr : Oe, x = e.rects.popper, m = e.elements[w ? b : y], O = Dn(pe(m) ? m : m.contextElement || ie(e.elements.popper), s, f, c), E = we(e.elements.reference), R = cr({
    reference: E,
    element: x,
    strategy: "absolute",
    placement: o
  }), B = at(Object.assign({}, x, R)), T = y === Oe ? B : E, A = {
    top: O.top - T.top + p.top,
    bottom: T.bottom - O.bottom + p.bottom,
    left: O.left - T.left + p.left,
    right: T.right - O.right + p.right
  }, C = e.modifiersData.offset;
  if (y === Oe && C) {
    var L = C[o];
    Object.keys(A).forEach(function(v) {
      var k = [q, N].indexOf(v) >= 0 ? 1 : -1, S = [j, N].indexOf(v) >= 0 ? "y" : "x";
      A[v] += L[S] * k;
    });
  }
  return A;
}
function Cn(e, t) {
  t === void 0 && (t = {});
  var r = t, n = r.placement, o = r.boundary, a = r.rootBoundary, c = r.padding, i = r.flipVariations, s = r.allowedAutoPlacements, u = s === void 0 ? tr : s, f = xe(n), h = f ? i ? Mt : Mt.filter(function(w) {
    return xe(w) === f;
  }) : De, y = h.filter(function(w) {
    return u.indexOf(w) >= 0;
  });
  y.length === 0 && (y = h);
  var g = y.reduce(function(w, d) {
    return w[d] = Ae(e, {
      placement: d,
      boundary: o,
      rootBoundary: a,
      padding: c
    })[K(d)], w;
  }, {});
  return Object.keys(g).sort(function(w, d) {
    return g[w] - g[d];
  });
}
function Ln(e) {
  if (K(e) === ct)
    return [];
  var t = Ne(e);
  return [Dt(e), t, Dt(t)];
}
function Wn(e) {
  var t = e.state, r = e.options, n = e.name;
  if (!t.modifiersData[n]._skip) {
    for (var o = r.mainAxis, a = o === void 0 ? !0 : o, c = r.altAxis, i = c === void 0 ? !0 : c, s = r.fallbackPlacements, u = r.padding, f = r.boundary, h = r.rootBoundary, y = r.altBoundary, g = r.flipVariations, w = g === void 0 ? !0 : g, d = r.allowedAutoPlacements, l = t.options.placement, p = K(l), b = p === l, x = s || (b || !w ? [Ne(l)] : Ln(l)), m = [l].concat(x).reduce(function(te, re) {
      return te.concat(K(re) === ct ? Cn(t, {
        placement: re,
        boundary: f,
        rootBoundary: h,
        padding: u,
        flipVariations: w,
        allowedAutoPlacements: d
      }) : re);
    }, []), O = t.rects.reference, E = t.rects.popper, R = /* @__PURE__ */ new Map(), B = !0, T = m[0], A = 0; A < m.length; A++) {
      var C = m[A], L = K(C), v = xe(C) === be, k = [j, N].indexOf(L) >= 0, S = k ? "width" : "height", M = Ae(t, {
        placement: C,
        boundary: f,
        rootBoundary: h,
        altBoundary: y,
        padding: u
      }), z = k ? v ? q : $ : v ? N : j;
      O[S] > E[S] && (z = Ne(z));
      var X = Ne(z), W = [];
      if (a && W.push(M[L] <= 0), i && W.push(M[z] <= 0, M[X] <= 0), W.every(function(te) {
        return te;
      })) {
        T = C, B = !1;
        break;
      }
      R.set(C, W);
    }
    if (B)
      for (var Y = w ? 3 : 1, V = function(re) {
        var Ee = m.find(function(Le) {
          var se = R.get(Le);
          if (se)
            return se.slice(0, re).every(function(Ge) {
              return Ge;
            });
        });
        if (Ee)
          return T = Ee, "break";
      }, ee = Y; ee > 0; ee--) {
        var ae = V(ee);
        if (ae === "break")
          break;
      }
    t.placement !== T && (t.modifiersData[n]._skip = !0, t.placement = T, t.reset = !0);
  }
}
const Fn = {
  name: "flip",
  enabled: !0,
  phase: "main",
  fn: Wn,
  requiresIfExists: ["offset"],
  data: {
    _skip: !1
  }
};
function Lt(e, t, r) {
  return r === void 0 && (r = {
    x: 0,
    y: 0
  }), {
    top: e.top - t.height - r.y,
    right: e.right - t.width + r.x,
    bottom: e.bottom - t.height + r.y,
    left: e.left - t.width - r.x
  };
}
function Wt(e) {
  return [j, q, N, $].some(function(t) {
    return e[t] >= 0;
  });
}
function jn(e) {
  var t = e.state, r = e.name, n = t.rects.reference, o = t.rects.popper, a = t.modifiersData.preventOverflow, c = Ae(t, {
    elementContext: "reference"
  }), i = Ae(t, {
    altBoundary: !0
  }), s = Lt(c, n), u = Lt(i, o, a), f = Wt(s), h = Wt(u);
  t.modifiersData[r] = {
    referenceClippingOffsets: s,
    popperEscapeOffsets: u,
    isReferenceHidden: f,
    hasPopperEscaped: h
  }, t.attributes.popper = Object.assign({}, t.attributes.popper, {
    "data-popper-reference-hidden": f,
    "data-popper-escaped": h
  });
}
const $n = {
  name: "hide",
  enabled: !0,
  phase: "main",
  requiresIfExists: ["preventOverflow"],
  fn: jn
};
function Hn(e, t, r) {
  var n = K(e), o = [$, j].indexOf(n) >= 0 ? -1 : 1, a = typeof r == "function" ? r(Object.assign({}, t, {
    placement: e
  })) : r, c = a[0], i = a[1];
  return c = c || 0, i = (i || 0) * o, [$, q].indexOf(n) >= 0 ? {
    x: i,
    y: c
  } : {
    x: c,
    y: i
  };
}
function In(e) {
  var t = e.state, r = e.options, n = e.name, o = r.offset, a = o === void 0 ? [0, 0] : o, c = tr.reduce(function(f, h) {
    return f[h] = Hn(h, t.rects, a), f;
  }, {}), i = c[t.placement], s = i.x, u = i.y;
  t.modifiersData.popperOffsets != null && (t.modifiersData.popperOffsets.x += s, t.modifiersData.popperOffsets.y += u), t.modifiersData[n] = c;
}
const Nn = {
  name: "offset",
  enabled: !0,
  phase: "main",
  requires: ["popperOffsets"],
  fn: In
};
function qn(e) {
  var t = e.state, r = e.name;
  t.modifiersData[r] = cr({
    reference: t.rects.reference,
    element: t.rects.popper,
    strategy: "absolute",
    placement: t.placement
  });
}
const Xn = {
  name: "popperOffsets",
  enabled: !0,
  phase: "read",
  fn: qn,
  data: {}
};
function Vn(e) {
  return e === "x" ? "y" : "x";
}
function Yn(e) {
  var t = e.state, r = e.options, n = e.name, o = r.mainAxis, a = o === void 0 ? !0 : o, c = r.altAxis, i = c === void 0 ? !1 : c, s = r.boundary, u = r.rootBoundary, f = r.altBoundary, h = r.padding, y = r.tether, g = y === void 0 ? !0 : y, w = r.tetherOffset, d = w === void 0 ? 0 : w, l = Ae(t, {
    boundary: s,
    rootBoundary: u,
    padding: h,
    altBoundary: f
  }), p = K(t.placement), b = xe(t.placement), x = !b, m = lt(p), O = Vn(m), E = t.modifiersData.popperOffsets, R = t.rects.reference, B = t.rects.popper, T = typeof d == "function" ? d(Object.assign({}, t.rects, {
    placement: t.placement
  })) : d, A = typeof T == "number" ? {
    mainAxis: T,
    altAxis: T
  } : Object.assign({
    mainAxis: 0,
    altAxis: 0
  }, T), C = t.modifiersData.offset ? t.modifiersData.offset[t.placement] : null, L = {
    x: 0,
    y: 0
  };
  if (E) {
    if (a) {
      var v, k = m === "y" ? j : $, S = m === "y" ? N : q, M = m === "y" ? "height" : "width", z = E[m], X = z + l[k], W = z - l[S], Y = g ? -B[M] / 2 : 0, V = b === be ? R[M] : B[M], ee = b === be ? -B[M] : -R[M], ae = t.elements.arrow, te = g && ae ? ut(ae) : {
        width: 0,
        height: 0
      }, re = t.modifiersData["arrow#persistent"] ? t.modifiersData["arrow#persistent"].padding : or(), Ee = re[k], Le = re[S], se = ke(0, R[M], te[M]), Ge = x ? R[M] / 2 - Y - se - Ee - A.mainAxis : V - se - Ee - A.mainAxis, br = x ? -R[M] / 2 + Y + se + Le + A.mainAxis : ee + se + Le + A.mainAxis, Je = t.elements.arrow && Ce(t.elements.arrow), yr = Je ? m === "y" ? Je.clientTop || 0 : Je.clientLeft || 0 : 0, gt = (v = C == null ? void 0 : C[m]) != null ? v : 0, wr = z + Ge - gt - yr, xr = z + br - gt, mt = ke(g ? Xe(X, wr) : X, z, g ? ue(W, xr) : W);
      E[m] = mt, L[m] = mt - z;
    }
    if (i) {
      var bt, Er = m === "x" ? j : $, Or = m === "x" ? N : q, ce = E[O], We = O === "y" ? "height" : "width", yt = ce + l[Er], wt = ce - l[Or], Qe = [j, $].indexOf(p) !== -1, xt = (bt = C == null ? void 0 : C[O]) != null ? bt : 0, Et = Qe ? yt : ce - R[We] - B[We] - xt + A.altAxis, Ot = Qe ? ce + R[We] + B[We] - xt - A.altAxis : wt, St = g && Qe ? gn(Et, ce, Ot) : ke(g ? Et : yt, ce, g ? Ot : wt);
      E[O] = St, L[O] = St - ce;
    }
    t.modifiersData[n] = L;
  }
}
const Un = {
  name: "preventOverflow",
  enabled: !0,
  phase: "main",
  fn: Yn,
  requiresIfExists: ["offset"]
};
function _n(e) {
  return {
    scrollLeft: e.scrollLeft,
    scrollTop: e.scrollTop
  };
}
function Kn(e) {
  return e === H(e) || !I(e) ? dt(e) : _n(e);
}
function Gn(e) {
  var t = e.getBoundingClientRect(), r = ye(t.width) / e.offsetWidth || 1, n = ye(t.height) / e.offsetHeight || 1;
  return r !== 1 || n !== 1;
}
function Jn(e, t, r) {
  r === void 0 && (r = !1);
  var n = I(t), o = I(t) && Gn(t), a = ie(t), c = we(e, o, r), i = {
    scrollLeft: 0,
    scrollTop: 0
  }, s = {
    x: 0,
    y: 0
  };
  return (n || !n && !r) && ((G(t) !== "body" || // https://github.com/popperjs/popper-core/issues/1078
  vt(a)) && (i = Kn(t)), I(t) ? (s = we(t, !0), s.x += t.clientLeft, s.y += t.clientTop) : a && (s.x = pt(a))), {
    x: c.left + i.scrollLeft - s.x,
    y: c.top + i.scrollTop - s.y,
    width: c.width,
    height: c.height
  };
}
function Qn(e) {
  var t = /* @__PURE__ */ new Map(), r = /* @__PURE__ */ new Set(), n = [];
  e.forEach(function(a) {
    t.set(a.name, a);
  });
  function o(a) {
    r.add(a.name);
    var c = [].concat(a.requires || [], a.requiresIfExists || []);
    c.forEach(function(i) {
      if (!r.has(i)) {
        var s = t.get(i);
        s && o(s);
      }
    }), n.push(a);
  }
  return e.forEach(function(a) {
    r.has(a.name) || o(a);
  }), n;
}
function Zn(e) {
  var t = Qn(e);
  return un.reduce(function(r, n) {
    return r.concat(t.filter(function(o) {
      return o.phase === n;
    }));
  }, []);
}
function eo(e) {
  var t;
  return function() {
    return t || (t = new Promise(function(r) {
      Promise.resolve().then(function() {
        t = void 0, r(e());
      });
    })), t;
  };
}
function to(e) {
  var t = e.reduce(function(r, n) {
    var o = r[n.name];
    return r[n.name] = o ? Object.assign({}, o, n, {
      options: Object.assign({}, o.options, n.options),
      data: Object.assign({}, o.data, n.data)
    }) : n, r;
  }, {});
  return Object.keys(t).map(function(r) {
    return t[r];
  });
}
var Ft = {
  placement: "bottom",
  modifiers: [],
  strategy: "absolute"
};
function jt() {
  for (var e = arguments.length, t = new Array(e), r = 0; r < e; r++)
    t[r] = arguments[r];
  return !t.some(function(n) {
    return !(n && typeof n.getBoundingClientRect == "function");
  });
}
function ro(e) {
  e === void 0 && (e = {});
  var t = e, r = t.defaultModifiers, n = r === void 0 ? [] : r, o = t.defaultOptions, a = o === void 0 ? Ft : o;
  return function(i, s, u) {
    u === void 0 && (u = a);
    var f = {
      placement: "bottom",
      orderedModifiers: [],
      options: Object.assign({}, Ft, a),
      modifiersData: {},
      elements: {
        reference: i,
        popper: s
      },
      attributes: {},
      styles: {}
    }, h = [], y = !1, g = {
      state: f,
      setOptions: function(p) {
        var b = typeof p == "function" ? p(f.options) : p;
        d(), f.options = Object.assign({}, a, f.options, b), f.scrollParents = {
          reference: pe(i) ? Re(i) : i.contextElement ? Re(i.contextElement) : [],
          popper: Re(s)
        };
        var x = Zn(to([].concat(n, f.options.modifiers)));
        return f.orderedModifiers = x.filter(function(m) {
          return m.enabled;
        }), w(), g.update();
      },
      // Sync update – it will always be executed, even if not necessary. This
      // is useful for low frequency updates where sync behavior simplifies the
      // logic.
      // For high frequency updates (e.g. `resize` and `scroll` events), always
      // prefer the async Popper#update method
      forceUpdate: function() {
        if (!y) {
          var p = f.elements, b = p.reference, x = p.popper;
          if (jt(b, x)) {
            f.rects = {
              reference: Jn(b, Ce(x), f.options.strategy === "fixed"),
              popper: ut(x)
            }, f.reset = !1, f.placement = f.options.placement, f.orderedModifiers.forEach(function(A) {
              return f.modifiersData[A.name] = Object.assign({}, A.data);
            });
            for (var m = 0; m < f.orderedModifiers.length; m++) {
              if (f.reset === !0) {
                f.reset = !1, m = -1;
                continue;
              }
              var O = f.orderedModifiers[m], E = O.fn, R = O.options, B = R === void 0 ? {} : R, T = O.name;
              typeof E == "function" && (f = E({
                state: f,
                options: B,
                name: T,
                instance: g
              }) || f);
            }
          }
        }
      },
      // Async and optimistically optimized update – it will not be executed if
      // not necessary (debounced to run at most once-per-tick)
      update: eo(function() {
        return new Promise(function(l) {
          g.forceUpdate(), l(f);
        });
      }),
      destroy: function() {
        d(), y = !0;
      }
    };
    if (!jt(i, s))
      return g;
    g.setOptions(u).then(function(l) {
      !y && u.onFirstUpdate && u.onFirstUpdate(l);
    });
    function w() {
      f.orderedModifiers.forEach(function(l) {
        var p = l.name, b = l.options, x = b === void 0 ? {} : b, m = l.effect;
        if (typeof m == "function") {
          var O = m({
            state: f,
            name: p,
            instance: g,
            options: x
          }), E = function() {
          };
          h.push(O || E);
        }
      });
    }
    function d() {
      h.forEach(function(l) {
        return l();
      }), h = [];
    }
    return g;
  };
}
var no = [Rn, Xn, Sn, pn, Nn, Fn, Un, wn, $n], oo = /* @__PURE__ */ ro({
  defaultModifiers: no
});
const $o = Object.freeze([
  "auto",
  "auto-start",
  "auto-end",
  "top",
  "top-start",
  "top-end",
  "bottom",
  "bottom-start",
  "bottom-end",
  "left",
  "left-start",
  "left-end",
  "right",
  "right-start",
  "right-end"
]);
function Ho(e) {
  const { placement: t, transfer: r, wrapper: n, isDrop: o = !1 } = e, a = e.reference ?? P(null), c = e.popper ?? P(null), i = P(""), s = {
    placement: t.value,
    modifiers: [
      {
        name: "preventOverflow",
        options: {
          rootBoundary: "window"
        }
      },
      {
        name: "computeStyles",
        options: {
          gpuAcceleration: !1
        }
      }
    ]
  };
  o && s.modifiers.push({
    name: "setTransformOrigin",
    enabled: !0,
    phase: "afterWrite",
    fn({ state: l }) {
      const p = io(l.placement);
      p && (l.elements.popper.style.transformOrigin = p);
    }
  }), e.offset && s.modifiers.push({
    name: "offset",
    options: {
      offset: e.offset
    }
  });
  let u = null;
  fe(t, (l) => {
    s.placement = l, u && u.setOptions({ placement: l });
  }), Rr(() => {
    n.value && c.value && (n.value.__transferElement = c.value);
  });
  let f = null;
  r && (fe(r, (l) => {
    g(l), h();
  }), g(r.value)), _e(() => {
    ze(y);
  }), ve(d);
  function h() {
    ze(() => {
      u && u.forceUpdate();
    });
  }
  function y() {
    d(), u && u.destroy(), w();
    const l = fe(a, w), p = fe(c, w);
    f = () => {
      l(), p();
    };
  }
  function g(l) {
    i.value = typeof l == "boolean" ? l ? "body" : "" : l;
  }
  function w() {
    a.value && c.value && (u && u.destroy(), u = oo(a.value, c.value, s), h());
  }
  function d() {
    u && u.destroy(), u = null, typeof f == "function" && (f(), f = null);
  }
  return {
    reference: a,
    popper: c,
    transferTo: i,
    updatePopper: h,
    createPopper: y,
    destroyPopper: d
  };
}
function io(e) {
  if (e !== "left" && e !== "right") {
    const [t, r] = e.split("-");
    return t === "bottom" || t !== "top" && r === "start" ? "center top" : "center bottom";
  }
}
var le = [], ao = function() {
  return le.some(function(e) {
    return e.activeTargets.length > 0;
  });
}, so = function() {
  return le.some(function(e) {
    return e.skippedTargets.length > 0;
  });
}, $t = "ResizeObserver loop completed with undelivered notifications.", co = function() {
  var e;
  typeof ErrorEvent == "function" ? e = new ErrorEvent("error", {
    message: $t
  }) : (e = document.createEvent("Event"), e.initEvent("error", !1, !1), e.message = $t), window.dispatchEvent(e);
}, Pe;
(function(e) {
  e.BORDER_BOX = "border-box", e.CONTENT_BOX = "content-box", e.DEVICE_PIXEL_CONTENT_BOX = "device-pixel-content-box";
})(Pe || (Pe = {}));
var de = function(e) {
  return Object.freeze(e);
}, fo = function() {
  function e(t, r) {
    this.inlineSize = t, this.blockSize = r, de(this);
  }
  return e;
}(), fr = function() {
  function e(t, r, n, o) {
    return this.x = t, this.y = r, this.width = n, this.height = o, this.top = this.y, this.left = this.x, this.bottom = this.top + this.height, this.right = this.left + this.width, de(this);
  }
  return e.prototype.toJSON = function() {
    var t = this, r = t.x, n = t.y, o = t.top, a = t.right, c = t.bottom, i = t.left, s = t.width, u = t.height;
    return { x: r, y: n, top: o, right: a, bottom: c, left: i, width: s, height: u };
  }, e.fromRect = function(t) {
    return new e(t.x, t.y, t.width, t.height);
  }, e;
}(), ht = function(e) {
  return e instanceof SVGElement && "getBBox" in e;
}, ur = function(e) {
  if (ht(e)) {
    var t = e.getBBox(), r = t.width, n = t.height;
    return !r && !n;
  }
  var o = e, a = o.offsetWidth, c = o.offsetHeight;
  return !(a || c || e.getClientRects().length);
}, Ht = function(e) {
  var t;
  if (e instanceof Element)
    return !0;
  var r = (t = e == null ? void 0 : e.ownerDocument) === null || t === void 0 ? void 0 : t.defaultView;
  return !!(r && e instanceof r.Element);
}, uo = function(e) {
  switch (e.tagName) {
    case "INPUT":
      if (e.type !== "image")
        break;
    case "VIDEO":
    case "AUDIO":
    case "EMBED":
    case "OBJECT":
    case "CANVAS":
    case "IFRAME":
    case "IMG":
      return !0;
  }
  return !1;
}, Te = typeof window < "u" ? window : {}, $e = /* @__PURE__ */ new WeakMap(), It = /auto|scroll/, lo = /^tb|vertical/, po = /msie|trident/i.test(Te.navigator && Te.navigator.userAgent), U = function(e) {
  return parseFloat(e || "0");
}, me = function(e, t, r) {
  return e === void 0 && (e = 0), t === void 0 && (t = 0), r === void 0 && (r = !1), new fo((r ? t : e) || 0, (r ? e : t) || 0);
}, Nt = de({
  devicePixelContentBoxSize: me(),
  borderBoxSize: me(),
  contentBoxSize: me(),
  contentRect: new fr(0, 0, 0, 0)
}), lr = function(e, t) {
  if (t === void 0 && (t = !1), $e.has(e) && !t)
    return $e.get(e);
  if (ur(e))
    return $e.set(e, Nt), Nt;
  var r = getComputedStyle(e), n = ht(e) && e.ownerSVGElement && e.getBBox(), o = !po && r.boxSizing === "border-box", a = lo.test(r.writingMode || ""), c = !n && It.test(r.overflowY || ""), i = !n && It.test(r.overflowX || ""), s = n ? 0 : U(r.paddingTop), u = n ? 0 : U(r.paddingRight), f = n ? 0 : U(r.paddingBottom), h = n ? 0 : U(r.paddingLeft), y = n ? 0 : U(r.borderTopWidth), g = n ? 0 : U(r.borderRightWidth), w = n ? 0 : U(r.borderBottomWidth), d = n ? 0 : U(r.borderLeftWidth), l = h + u, p = s + f, b = d + g, x = y + w, m = i ? e.offsetHeight - x - e.clientHeight : 0, O = c ? e.offsetWidth - b - e.clientWidth : 0, E = o ? l + b : 0, R = o ? p + x : 0, B = n ? n.width : U(r.width) - E - O, T = n ? n.height : U(r.height) - R - m, A = B + l + O + b, C = T + p + m + x, L = de({
    devicePixelContentBoxSize: me(Math.round(B * devicePixelRatio), Math.round(T * devicePixelRatio), a),
    borderBoxSize: me(A, C, a),
    contentBoxSize: me(B, T, a),
    contentRect: new fr(h, s, B, T)
  });
  return $e.set(e, L), L;
}, dr = function(e, t, r) {
  var n = lr(e, r), o = n.borderBoxSize, a = n.contentBoxSize, c = n.devicePixelContentBoxSize;
  switch (t) {
    case Pe.DEVICE_PIXEL_CONTENT_BOX:
      return c;
    case Pe.BORDER_BOX:
      return o;
    default:
      return a;
  }
}, vo = function() {
  function e(t) {
    var r = lr(t);
    this.target = t, this.contentRect = r.contentRect, this.borderBoxSize = de([r.borderBoxSize]), this.contentBoxSize = de([r.contentBoxSize]), this.devicePixelContentBoxSize = de([r.devicePixelContentBoxSize]);
  }
  return e;
}(), pr = function(e) {
  if (ur(e))
    return 1 / 0;
  for (var t = 0, r = e.parentNode; r; )
    t += 1, r = r.parentNode;
  return t;
}, ho = function() {
  var e = 1 / 0, t = [];
  le.forEach(function(c) {
    if (c.activeTargets.length !== 0) {
      var i = [];
      c.activeTargets.forEach(function(u) {
        var f = new vo(u.target), h = pr(u.target);
        i.push(f), u.lastReportedSize = dr(u.target, u.observedBox), h < e && (e = h);
      }), t.push(function() {
        c.callback.call(c.observer, i, c.observer);
      }), c.activeTargets.splice(0, c.activeTargets.length);
    }
  });
  for (var r = 0, n = t; r < n.length; r++) {
    var o = n[r];
    o();
  }
  return e;
}, qt = function(e) {
  le.forEach(function(r) {
    r.activeTargets.splice(0, r.activeTargets.length), r.skippedTargets.splice(0, r.skippedTargets.length), r.observationTargets.forEach(function(o) {
      o.isActive() && (pr(o.target) > e ? r.activeTargets.push(o) : r.skippedTargets.push(o));
    });
  });
}, go = function() {
  var e = 0;
  for (qt(e); ao(); )
    e = ho(), qt(e);
  return so() && co(), e > 0;
}, tt, vr = [], mo = function() {
  return vr.splice(0).forEach(function(e) {
    return e();
  });
}, bo = function(e) {
  if (!tt) {
    var t = 0, r = document.createTextNode(""), n = { characterData: !0 };
    new MutationObserver(function() {
      return mo();
    }).observe(r, n), tt = function() {
      r.textContent = "".concat(t ? t-- : t++);
    };
  }
  vr.push(e), tt();
}, yo = function(e) {
  bo(function() {
    requestAnimationFrame(e);
  });
}, qe = 0, wo = function() {
  return !!qe;
}, xo = 250, Eo = { attributes: !0, characterData: !0, childList: !0, subtree: !0 }, Xt = [
  "resize",
  "load",
  "transitionend",
  "animationend",
  "animationstart",
  "animationiteration",
  "keyup",
  "keydown",
  "mouseup",
  "mousedown",
  "mouseover",
  "mouseout",
  "blur",
  "focus"
], Vt = function(e) {
  return e === void 0 && (e = 0), Date.now() + e;
}, rt = !1, Oo = function() {
  function e() {
    var t = this;
    this.stopped = !0, this.listener = function() {
      return t.schedule();
    };
  }
  return e.prototype.run = function(t) {
    var r = this;
    if (t === void 0 && (t = xo), !rt) {
      rt = !0;
      var n = Vt(t);
      yo(function() {
        var o = !1;
        try {
          o = go();
        } finally {
          if (rt = !1, t = n - Vt(), !wo())
            return;
          o ? r.run(1e3) : t > 0 ? r.run(t) : r.start();
        }
      });
    }
  }, e.prototype.schedule = function() {
    this.stop(), this.run();
  }, e.prototype.observe = function() {
    var t = this, r = function() {
      return t.observer && t.observer.observe(document.body, Eo);
    };
    document.body ? r() : Te.addEventListener("DOMContentLoaded", r);
  }, e.prototype.start = function() {
    var t = this;
    this.stopped && (this.stopped = !1, this.observer = new MutationObserver(this.listener), this.observe(), Xt.forEach(function(r) {
      return Te.addEventListener(r, t.listener, !0);
    }));
  }, e.prototype.stop = function() {
    var t = this;
    this.stopped || (this.observer && this.observer.disconnect(), Xt.forEach(function(r) {
      return Te.removeEventListener(r, t.listener, !0);
    }), this.stopped = !0);
  }, e;
}(), st = new Oo(), Yt = function(e) {
  !qe && e > 0 && st.start(), qe += e, !qe && st.stop();
}, So = function(e) {
  return !ht(e) && !uo(e) && getComputedStyle(e).display === "inline";
}, ko = function() {
  function e(t, r) {
    this.target = t, this.observedBox = r || Pe.CONTENT_BOX, this.lastReportedSize = {
      inlineSize: 0,
      blockSize: 0
    };
  }
  return e.prototype.isActive = function() {
    var t = dr(this.target, this.observedBox, !0);
    return So(this.target) && (this.lastReportedSize = t), this.lastReportedSize.inlineSize !== t.inlineSize || this.lastReportedSize.blockSize !== t.blockSize;
  }, e;
}(), Ro = function() {
  function e(t, r) {
    this.activeTargets = [], this.skippedTargets = [], this.observationTargets = [], this.observer = t, this.callback = r;
  }
  return e;
}(), He = /* @__PURE__ */ new WeakMap(), Ut = function(e, t) {
  for (var r = 0; r < e.length; r += 1)
    if (e[r].target === t)
      return r;
  return -1;
}, Ie = function() {
  function e() {
  }
  return e.connect = function(t, r) {
    var n = new Ro(t, r);
    He.set(t, n);
  }, e.observe = function(t, r, n) {
    var o = He.get(t), a = o.observationTargets.length === 0;
    Ut(o.observationTargets, r) < 0 && (a && le.push(o), o.observationTargets.push(new ko(r, n && n.box)), Yt(1), st.schedule());
  }, e.unobserve = function(t, r) {
    var n = He.get(t), o = Ut(n.observationTargets, r), a = n.observationTargets.length === 1;
    o >= 0 && (a && le.splice(le.indexOf(n), 1), n.observationTargets.splice(o, 1), Yt(-1));
  }, e.disconnect = function(t) {
    var r = this, n = He.get(t);
    n.observationTargets.slice().forEach(function(o) {
      return r.unobserve(t, o.target);
    }), n.activeTargets.splice(0, n.activeTargets.length);
  }, e;
}(), _t = function() {
  function e(t) {
    if (arguments.length === 0)
      throw new TypeError("Failed to construct 'ResizeObserver': 1 argument required, but only 0 present.");
    if (typeof t != "function")
      throw new TypeError("Failed to construct 'ResizeObserver': The callback provided as parameter 1 is not a function.");
    Ie.connect(this, t);
  }
  return e.prototype.observe = function(t, r) {
    if (arguments.length === 0)
      throw new TypeError("Failed to execute 'observe' on 'ResizeObserver': 1 argument required, but only 0 present.");
    if (!Ht(t))
      throw new TypeError("Failed to execute 'observe' on 'ResizeObserver': parameter 1 is not of type 'Element");
    Ie.observe(this, t, r);
  }, e.prototype.unobserve = function(t) {
    if (arguments.length === 0)
      throw new TypeError("Failed to execute 'unobserve' on 'ResizeObserver': 1 argument required, but only 0 present.");
    if (!Ht(t))
      throw new TypeError("Failed to execute 'unobserve' on 'ResizeObserver': parameter 1 is not of type 'Element");
    Ie.unobserve(this, t);
  }, e.prototype.disconnect = function() {
    Ie.disconnect(this);
  }, e.toString = function() {
    return "function ResizeObserver () { [polyfill code] }";
  }, e;
}();
const Ve = /* @__PURE__ */ new WeakMap();
function To(e) {
  for (let t = 0, r = e.length; t < r; ++t) {
    const n = e[t], o = Ve.get(n.target);
    typeof o == "function" && o(n);
  }
}
const hr = new (oe && window.ResizeObserver || _t)(
  To
);
function gr(e, t) {
  Ve.set(e, t), hr.observe(e);
}
function mr(e) {
  Ve.has(e) && (hr.unobserve(e), Ve.delete(e));
}
function Io() {
  return {
    observeResize: gr,
    unobserveResize: mr
  };
}
function No() {
  const e = {};
  return ve(() => {
    Object.keys(e).forEach((t) => {
      clearTimeout(e[t]);
    });
  }), { timer: e };
}
function qo() {
  const e = {};
  return ve(() => {
    Object.keys(e).forEach((t) => {
      clearInterval(e[t]);
    });
  }), { timer: e };
}
function Xo(e, t, r = 250) {
  let n;
  function o() {
    e.value === "hover" && (clearTimeout(n), n = setTimeout(() => {
      t.value = !0;
    }, r));
  }
  function a() {
    e.value === "hover" && (clearTimeout(n), n = setTimeout(() => {
      t.value = !1;
    }, r));
  }
  function c() {
    e.value === "click" && (t.value = !t.value);
  }
  return {
    handleTriggerEnter: o,
    handleTriggerLeave: a,
    handleTriggerClick: c
  };
}
function Vo(e) {
  const {
    items: t,
    itemSize: r,
    itemFixed: n,
    idKey: o,
    defaultKeyAt: a,
    bufferSize: c = P(5),
    wrapper: i = P(null),
    autoResize: s = !0
  } = e, u = _(() => {
    const v = D(t), k = v.length, S = D(o), M = /* @__PURE__ */ new Map();
    for (let z = 0; z < k; ++z)
      M.set(v[z][S], z);
    return M;
  }), f = P(0), h = /* @__PURE__ */ new Map(), y = _(() => {
    const v = D(t), k = v.length, S = D(o), M = Br(k, D(r));
    for (let z = 0; z < k; ++z) {
      const X = v[z][S], W = h.get(X);
      W && M.add(z, W);
    }
    return M;
  }), g = P(0), w = P(0), d = _(() => Math.max(
    y.value.boundIndex(w.value) - Math.max(D(c), 0),
    0
  )), l = _(() => {
    if (!f.value || f.value < 0)
      return [];
    const v = D(t), k = Math.min(
      y.value.boundIndex(w.value + f.value) + 1 + Math.max(D(c), 0),
      v.length
    );
    return v.slice(d.value, k);
  }), p = _(() => {
    g.value;
    const v = y.value.sum(), k = D(n);
    return {
      height: k ? `${v}px` : void 0,
      minHeight: k ? void 0 : `${v}px`,
      boxSizing: "content-box"
    };
  }), b = _(() => (g.value, {
    transform: `translate3d(0, ${y.value.sum(d.value)}px, 0)`
  }));
  _e(() => {
    ze(() => {
      const v = D(i);
      s && v && gr(v, O), nt(a) && T(a);
    });
  }), s && ve(() => {
    const v = D(i);
    v && mr(v);
  });
  function x() {
    const v = D(i);
    v && (w.value = v.scrollTop);
  }
  function m() {
    Hr(x);
  }
  function O(v) {
    zt(v.target) || v.contentRect.height === f.value || (f.value = v.contentRect.height);
  }
  function E(v, k) {
    var ee, ae;
    if (D(n))
      return;
    const S = u.value.get(v), M = y.value, z = M.get(S), X = ((ae = (ee = k.borderBoxSize) == null ? void 0 : ee[0]) == null ? void 0 : ae.blockSize) ?? k.contentRect.height;
    if (X === z)
      return;
    if (zt(k.target)) {
      z && (M.add(S, -z), g.value++);
      return;
    }
    const W = X - D(r), Y = X - z;
    if (W ? h.set(v, W) : h.delete(v), !Y)
      return;
    M.add(S, Y), g.value++;
    const V = D(i);
    if (V) {
      const te = M.sum(S);
      V.scrollTop > te && V.scrollBy(0, Y), w.value = V.scrollTop;
    }
  }
  function R(v, k) {
    const S = D(i);
    S && S.scrollTo({
      behavior: k,
      top: v,
      left: 0
    });
  }
  function B(v, k) {
    const S = D(i);
    S && S.scrollBy({
      behavior: k,
      top: v,
      left: 0
    });
  }
  function T(v, k) {
    const S = u.value.get(v);
    S != null && A(S, k);
  }
  function A(v, k) {
    const S = D(i);
    S && S.scrollTo({
      behavior: k,
      top: y.value.sum(v),
      left: 0
    });
  }
  function C(v, k) {
    const S = D(i);
    if (!S)
      return;
    const M = y.value, z = S.scrollTop;
    if (M.sum(v) < z) {
      A(v, k);
      return;
    }
    const W = S.offsetHeight, Y = z + W, V = M.sum(v + 1);
    V > Y && R(V - W, k);
  }
  function L(v, k) {
    const S = u.value.get(v);
    S != null && C(S, k);
  }
  return {
    wrapper: i,
    indexMap: u,
    heightTree: y,
    startIndex: d,
    scrollOffset: w,
    visibleItems: l,
    listStyle: p,
    itemsStyle: b,
    handleScroll: m,
    handleResize: O,
    handleItemResize: E,
    scrollTo: R,
    scrollBy: B,
    scrollToKey: T,
    scrollToIndex: A,
    ensureIndexInView: C,
    ensureKeyInView: L
  };
}
function Yo(e, t, r) {
  for (const n of t)
    if (e[n])
      return (o) => Tr(e, n, o);
  return r || null;
}
export {
  he as CLICK_OUTSIDE,
  Yo as createSlotRender,
  zt as isHiddenElement,
  gr as observeResize,
  $o as placementWhileList,
  Jt as queryOutsideHiddenElement,
  mr as unobserveResize,
  Bo as useClickOutside,
  Mo as useDisplay,
  Ao as useFullScreen,
  Po as useHover,
  Do as useIntersection,
  Q as useListener,
  Co as useManualRef,
  Lo as useModifier,
  Wo as useMounted,
  Fo as useMoving,
  jo as usePointer,
  Ho as usePopper,
  Io as useResize,
  qo as useSetInterval,
  No as useSetTimeout,
  Xo as useTriggerHandler,
  Vo as useVirtual
};
