// @tanstack/vue-table/build/lib/index.js
/**
 * vue-table
 *
 * Copyright (c) TanStack
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.md file in the root directory of this source tree.
 *
 * @license MIT
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tableCore = require('@tanstack/table-core');
var vue = require('vue');
var mergeProxy = require('./merge-proxy.js');

const FlexRender = vue.defineComponent({
  props: ['render', 'props'],
  setup: props => {
    return () => {
      if (typeof props.render === 'function' || typeof props.render === 'object') {
        return vue.h(props.render, props.props);
      }
      return props.render;
    };
  }
});
function useVueTable(options) {
  const resolvedOptions = mergeProxy.mergeProxy({
    state: {},
    // Dummy state
    onStateChange: () => {},
    // noop
    renderFallbackValue: null,
    mergeOptions(defaultOptions, options) {
      return mergeProxy.mergeProxy(defaultOptions, options);
    }
  }, options);
  const table = tableCore.createTable(resolvedOptions);
  // can't use `reactive` because update needs to be immutable
  const state = vue.ref(table.initialState);
  vue.watchEffect(() => {
    table.setOptions(prev => {
      var _options$state;
      const stateProxy = new Proxy({}, {
        get: (_, prop) => state.value[prop]
      });
      return mergeProxy.mergeProxy(prev, options, {
        // merge the initialState and `options.state`
        // create a new proxy on each `setOptions` call
        // and get the value from state on each property access
        state: mergeProxy.mergeProxy(stateProxy, (_options$state = options.state) != null ? _options$state : {}),
        // Similarly, we'll maintain both our internal state and any user-provided
        // state.
        onStateChange: updater => {
          if (updater instanceof Function) {
            state.value = updater(state.value);
          } else {
            state.value = updater;
          }
          options.onStateChange == null || options.onStateChange(updater);
        }
      });
    });
  });
  return table;
}

exports.FlexRender = FlexRender;
exports.useVueTable = useVueTable;
Object.keys(tableCore).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return tableCore[k]; }
  });
});
